src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs

```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+                        return _compiledLibraries[context.RootProject.Identity.Name] = true;
+                    }
+                }
+                else
+                {
+                    // Write RSP file for the next time
+                    File.WriteAllLines(rsp, allArgs);
+
+                    Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}:  Previously compiled, skipping dynamic compilation."));
+                    return _compiledLibraries[context.RootProject.Identity.Name] = true;
+                }
+            }
+
+            Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}: Dynamic compiling for {context.TargetFramework.DotNetFrameworkName}"));
+
+            // Write the dependencies file
+            if (dependencyContext != null)
+            {
+                var writer = new DependencyContextWriter();
+                using (var fileStream = File.Create(depsJsonFile))
+                {
+                    writer.Write(dependencyContext, fileStream);
+                }
+            }
+
+            // Write assembly info and RSP files
+            File.WriteAllText(assemblyInfo, AssemblyInfoFileGenerator.GenerateCSharp(assemblyInfoOptions, sourceFiles));
+            File.WriteAllLines(rsp, allArgs);
+
+            // Locate runtime config files
+            var runtimeConfigPath = Path.Combine(runtimeDirectory, EntryAssembly.GetName().Name + FileNameSuffixes.RuntimeConfigJson);
+            var cscRuntimeConfigPath =  Path.Combine(runtimeDirectory, "csc" + FileNameSuffixes.RuntimeConfigJson);
+
+            // Automatically create the csc runtime config file
+            if (File.Exists(runtimeConfigPath) && (!File.Exists(cscRuntimeConfigPath)
+                || File.GetLastWriteTimeUtc(runtimeConfigPath) > File.GetLastWriteTimeUtc(cscRuntimeConfigPath)))
+            {
+                lock (_syncLock)
+                {
+                    File.Copy(runtimeConfigPath, cscRuntimeConfigPath, true);
+                }
+            }
 
             // Execute CSC!
-            var result = RunCsc(allArgs.ToArray())
-                .WorkingDirectory(Directory.GetCurrentDirectory())
-                .ForwardStdErr()
-                .ForwardStdOut()
+            var result = Command.Create("csc.dll", new string[] { $"-noconfig", "@" + $"{rsp}" })
+                .WorkingDirectory(runtimeDirectory)
+                .OnErrorLine(line => Diagnostics.Add(line))
+                .OnOutputLine(line => Diagnostics.Add(line))
                 .Execute();
 
-            return result.ExitCode == 0;
+            Debug.WriteLine(String.Empty);
+
+            if (result.ExitCode == 0 && !Diagnostics.Any())
+            {
+                Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}: Dynamic compilation succeeded."));
+                Debug.WriteLine($"0 Warning(s)");
+                Debug.WriteLine($"0 Error(s)");
```

Comment:
Oh yes but since i use parallel compilation, here i needed to specify again the project name with the compilation result message.